virtualbox: Unable to compile VirtualBox kernel modules with kernel 5.10

**Describe the bug**
VirtualBox kernel modules are not able to compile with `kernel_latest`.

`configuration.nix`
```
...
boot.kernelPackages = pkgs.linuxPackages_latest;
...
virtualisation.virtualbox.host.enable = true;
...
```

```
 asbachb@nixo-t14s  ~  sudo nixos-rebuild switch
building Nix...
building the system configuration...
these derivations will be built:
  /nix/store/q7fyvhkz53pryr8bxh6r6zdz87lz479r-system-path.drv
  /nix/store/dbylza6h4dr36i39y2yx2g52y44dfrcc-dbus-1.drv
  /nix/store/0nhc7z7zda2amhlzlbi0a7svp8lckskk-unit-dbus.service.drv
  /nix/store/jmjh19xfhsjwm5lbdc3s1h967zrbwhbh-unit-dbus.service.drv
  /nix/store/qxp8zmc88irr3n5pwd0w5ngkg7qnqmr4-unit-systemd-fsck-.service.drv
  /nix/store/z8fdcsf3dwj5icwxf648789x6nws4m9y-unit-polkit.service.drv
  /nix/store/13krbx2g227viybw4498g31hxfvaw02b-system-units.drv
  /nix/store/far87ahiirm9i2dghqzi974fsqw0f8m6-virtualbox-modules-6.1.16-5.10.2.drv
  /nix/store/jww9dbspydkv86fnclijrxa02hbcn0i9-linux-5.10.2-modules.drv
  /nix/store/w25ppabibq4a795730xnqlb0hmrj83lg-linux-5.10.2-modules-shrunk.drv
  /nix/store/4s01w1i4cgjw6yd9rrm3nyn9r9b0wads-stage-1-init.sh.drv
  /nix/store/n4nr93sjv8a9jcnpih71qh9gc5zlgpjx-user-units.drv
  /nix/store/94pagcnsfgyyvikkhiwm69jbvd1w9zcy-etc.drv
  /nix/store/fhbb3q2jsgrx66azr1qhb4nfi0svqp8v-initrd-linux-5.10.2.drv
  /nix/store/qa56gv400bln95vaifnfc4rwjsf2w5rn-kernel-modules.drv
  /nix/store/x8hjrmi4w8dcdiyr7sjv3apjzr7i0bqc-nixos-system-nixo-t14s-21.03pre259711.be0b453d7c7.drv
building '/nix/store/q7fyvhkz53pryr8bxh6r6zdz87lz479r-system-path.drv'...
building '/nix/store/far87ahiirm9i2dghqzi974fsqw0f8m6-virtualbox-modules-6.1.16-5.10.2.drv'...
unpacking sources
unpacking source archive /nix/store/q12mjnim3j4mx009816y2j2rn8lbc5px-virtualbox-6.1.16-modsrc
source root is virtualbox-6.1.16-modsrc
patching sources
configuring
no configure script, doing nothing
building
build flags: -j16 -l16 SHELL=/nix/store/4l7wsi6h6283194r6fqy1731qxlagq62-bash-4.4-p23/bin/bash INSTALL_MOD_PATH=\$\(out\)
=== Building 'vboxdrv' module ===
make[1]: Entering directory '/build/virtualbox-6.1.16-modsrc/vboxdrv'
make V= CONFIG_MODULE_SIG= CONFIG_MODULE_SIG_ALL= -C /nix/store/vfjkakb00wa967gphgbrw72lhqlhrh6j-linux-5.10.2-dev/lib/modules/5.10.2/build M=/build/virtualbox-6.1.16-modsrc/vboxdrv SRCROOT=/build/virtualbox-6.1.16-modsrc/vboxdrv -j16 modules
make[2]: Entering directory '/build/virtualbox-6.1.16-modsrc/vboxdrv'
make[2]: warning: -j16 forced in submake: resetting jobserver mode.
collision between `/nix/store/3ln42pm0g5iak1iddi8wj6qp2l31i2hx-nix-bash-completions-0.6.8/share/bash-completion/completions/nixos-container' and `/nix/store/gmgddv8s3qnhx9d940rpl9gczawz40gy-nixos-container/share/bash-completion/completions/nixos-container'
created 8507 symlinks in user environment
building '/nix/store/dbylza6h4dr36i39y2yx2g52y44dfrcc-dbus-1.drv'...
building '/nix/store/z8fdcsf3dwj5icwxf648789x6nws4m9y-unit-polkit.service.drv'...
building '/nix/store/qxp8zmc88irr3n5pwd0w5ngkg7qnqmr4-unit-systemd-fsck-.service.drv'...
building '/nix/store/0nhc7z7zda2amhlzlbi0a7svp8lckskk-unit-dbus.service.drv'...
building '/nix/store/jmjh19xfhsjwm5lbdc3s1h967zrbwhbh-unit-dbus.service.drv'...
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/linux/SUPDrv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/SUPDrv.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/SUPDrvGip.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/SUPDrvSem.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/SUPDrvTracer.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/SUPLibAll.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/alloc-r0drv.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/initterm-r0drv.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/memobj-r0drv.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/mpnotification-r0drv.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/powernotification-r0drv.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/assert-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/alloc-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/initterm-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/memobj-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/memuserkernel-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/mp-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/mpnotification-r0drv-linux.o
building '/nix/store/n4nr93sjv8a9jcnpih71qh9gc5zlgpjx-user-units.drv'...
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/process-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/rtStrFormatKernelAddress-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/semevent-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/semeventmulti-r0drv-linux.o
/build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/memobj-r0drv-linux.c: In function 'rtR0MemObjLinuxVMap':
/build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/memobj-r0drv-linux.c:560:34: error: implicit declaration of function 'alloc_vm_area' [-Werror=implicit-function-declaration]
  560 |                 pMemLnx->pArea = alloc_vm_area(pMemLnx->Core.cb, papPtes); /* Note! pArea->nr_pages is not set. */
      |                                  ^~~~~~~~~~~~~
/build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/memobj-r0drv-linux.c:560:32: warning: assignment to 'struct vm_struct *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  560 |                 pMemLnx->pArea = alloc_vm_area(pMemLnx->Core.cb, papPtes); /* Note! pArea->nr_pages is not set. */
      |                                ^
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/semfastmutex-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/semmutex-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/spinlock-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/thread-r0drv-linux.o
/build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/memuserkernel-r0drv-linux.o: warning: objtool: VBoxHost_RTR0MemKernelCopyFrom()+0x13: redundant CLD
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/thread2-r0drv-linux.o
building '/nix/store/13krbx2g227viybw4498g31hxfvaw02b-system-units.drv'...
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/threadctxhooks-r0drv-linux.o
/build/virtualbox-6.1.16-modsrc/vboxdrv/SUPDrvTracer.o: warning: objtool: .text+0x7: indirect jump found in RETPOLINE build
/build/virtualbox-6.1.16-modsrc/vboxdrv/SUPDrvTracer.o: warning: objtool: supdrvTracerProbeFireStub() is missing an ELF size annotation
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/time-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/timer-r0drv-linux.o
  CC [M]  /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/generic/semspinmutex-r0drv-generic.o
cc1: some warnings being treated as errors
make[4]: *** [/nix/store/vfjkakb00wa967gphgbrw72lhqlhrh6j-linux-5.10.2-dev/lib/modules/5.10.2/source/scripts/Makefile.build:279: /build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/memobj-r0drv-linux.o] Error 1
make[4]: *** Waiting for unfinished jobs....
/build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/threadctxhooks-r0drv-linux.o: warning: objtool: rtThreadCtxHooksLnxSchedOut()+0x1f: call to __x86_indirect_thunk_rax() with UACCESS enabled
/build/virtualbox-6.1.16-modsrc/vboxdrv/r0drv/linux/threadctxhooks-r0drv-linux.o: warning: objtool: rtThreadCtxHooksLnxSchedIn()+0x29: call to __x86_indirect_thunk_rax() with UACCESS enabled
make[3]: *** [/nix/store/vfjkakb00wa967gphgbrw72lhqlhrh6j-linux-5.10.2-dev/lib/modules/5.10.2/source/Makefile:1805: /build/virtualbox-6.1.16-modsrc/vboxdrv] Error 2
make[2]: *** [/nix/store/vfjkakb00wa967gphgbrw72lhqlhrh6j-linux-5.10.2-dev/lib/modules/5.10.2/source/Makefile:185: __sub-make] Error 2
make[2]: Leaving directory '/nix/store/vfjkakb00wa967gphgbrw72lhqlhrh6j-linux-5.10.2-dev/lib/modules/5.10.2/build'
make[1]: *** [/build/virtualbox-6.1.16-modsrc/vboxdrv/Makefile-footer.gmk:117: vboxdrv] Error 2
make[1]: Leaving directory '/build/virtualbox-6.1.16-modsrc/vboxdrv'
make: *** [Makefile:60: vboxdrv] Error 2
builder for '/nix/store/far87ahiirm9i2dghqzi974fsqw0f8m6-virtualbox-modules-6.1.16-5.10.2.drv' failed with exit code 2
cannot build derivation '/nix/store/qa56gv400bln95vaifnfc4rwjsf2w5rn-kernel-modules.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/jww9dbspydkv86fnclijrxa02hbcn0i9-linux-5.10.2-modules.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/x8hjrmi4w8dcdiyr7sjv3apjzr7i0bqc-nixos-system-nixo-t14s-21.03pre259711.be0b453d7c7.drv': 1 dependencies couldn't be built
error: build of '/nix/store/x8hjrmi4w8dcdiyr7sjv3apjzr7i0bqc-nixos-system-nixo-t14s-21.03pre259711.be0b453d7c7.drv' failed

```

**Notify maintainers**
-

**Metadata**
```
 ✘ asbachb@nixo-t14s  ~  nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.2, NixOS, 21.03pre259711.be0b453d7c7 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-21.03pre259711.be0b453d7c7, nixos-unstable"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

/cc @Magicloud @Ma27 @flokli @B4dM4n @bryanasdev000

--- a/pkgs/applications/virtualization/virtualbox/default.nix
+++ b/pkgs/applications/virtualization/virtualbox/default.nix
@@ -23,7 +23,7 @@ let
   buildType = "release";
   # Use maintainers/scripts/update.nix to update the version and all related hashes or
   # change the hashes in extpack.nix and guest-additions/default.nix as well manually.
-  version = "6.1.16";
+  version = "6.1.18";
 
   iasl' = iasl.overrideAttrs (old: rec {
     inherit (old) pname;
@@ -40,7 +40,7 @@ in stdenv.mkDerivation {
 
   src = fetchurl {
     url = "https://download.virtualbox.org/virtualbox/${version}/VirtualBox-${version}.tar.bz2";
-    sha256 = "49c1990da16d8a3d5bda8cdb961ec8195a901e67e4c79aea44c1521a5fc2f9f1";
+    sha256 = "108d42b9b391b7a332a33df1662cf7b0e9d9a80f3079d16288d8b9487f427d40";
   };
 
   outputs = [ "out" "modsrc" ];

--- a/pkgs/applications/virtualization/virtualbox/extpack.nix
+++ b/pkgs/applications/virtualization/virtualbox/extpack.nix
@@ -12,7 +12,7 @@ fetchurl rec {
     # Manually sha256sum the extensionPack file, must be hex!
     # Thus do not use `nix-prefetch-url` but instead plain old `sha256sum`.
     # Checksums can also be found at https://www.virtualbox.org/download/hashes/${version}/SHA256SUMS
-    let value = "9802482b77b95a954cb5111793da10d009009a4e9a9c4eaa4bd1ae5dafe9db46";
+    let value = "d609e35accff5c0819ca9be47de302abf094dc1b6d4c54da8fdda639671f267e";
     in assert (builtins.stringLength value) == 64; value;
 
   meta = {

--- a/pkgs/applications/virtualization/virtualbox/guest-additions/default.nix
+++ b/pkgs/applications/virtualization/virtualbox/guest-additions/default.nix
@@ -27,7 +27,7 @@ in stdenv.mkDerivation rec {
 
   src = fetchurl {
     url = "http://download.virtualbox.org/virtualbox/${version}/VBoxGuestAdditions_${version}.iso";
-    sha256 = "88db771a5efd7c048228e5c1e0b8fba56542e9d8c1b75f7af5b0c4cf334f0584";
+    sha256 = "904432eb331d7ae517afaa4e4304e6492b7947b46ecb8267de7ef792c4921b4c";
   };
 
   KERN_DIR = "${kernel.dev}/lib/modules/${kernel.modDirVersion}/build";

